cmd/gcp-controller-manager/dpwi/nodesyncer/nodesyncer.go

```diff
@@ -0,0 +1,263 @@
+	if key == "" {
+		return nil
+	}
+	pods, err := eh.podIndexer.ByIndex(indexByNode, key)
+	if err != nil {
+		return fmt.Errorf("failed to get pods on the node %q: %v", key, err)
+	}
+
+	if key == "" {
+		return nil
+	}
+	gsaSet := make(map[string]int)
+	for _, p := range pods {
+		pod, ok := p.(*core.Pod)
+		if !ok {
+			return fmt.Errorf("invalid pod object from key %q", p)
+		}
+
+		ksa := serviceaccounts.ServiceAccount{
+			Namespace: pod.ObjectMeta.Namespace,
+			Name:      pod.Spec.ServiceAccountName,
+		}
+		gsa, err := eh.verifier.VerifiedGSA(ksa)
+		if err != nil {
+			return fmt.Errorf("failed to get verified GSA for ksa %v: %v", ksa, err)
+		}
+		if gsa == "" {
+			continue
+		}
+		gsaSet[string(gsa)]++
+	}
+	var curGSAs []string
+	for k := range gsaSet {
+		curGSAs = append(curGSAs, k)
+	}
+	zone, err := eh.zones.zoneByNode(key)
+	if err != nil {
+		return err
+	}
+	return eh.nodeMap.getNodeGSAs(key).sync(
+		ctx,
+		eh.hms,
+		zone,
+		curGSAs,
+	)
+}
+
+type nodeMap struct {
+	sync.Mutex
+	m map[string]*nodeGSAs
+}
+
+func (nm *nodeMap) getNodeGSAs(node string) *nodeGSAs {
+	nm.Lock()
+	defer nm.Unlock()
+	if v, ok := nm.m[node]; ok {
+		return v
+	}
+	entry := &nodeGSAs{
+		node: node,
+	}
+	nm.m[node] = entry
+	return entry
+}
+
+type nodeGSAs struct {
+	sync.Mutex
+	lastGSAs []string
+	lastSent time.Time
+	node     string
+}
+
+func (ng *nodeGSAs) sync(ctx context.Context, client *hms.Client, zone string, curGSAs []string) error {
+	ng.Lock()
+	defer ng.Unlock()
+	sort.Strings(curGSAs)
+	if equal := reflect.DeepEqual(ng.lastGSAs, curGSAs); equal {
+		klog.Infof("Don't sync for node %q as the GSA list is the same", ng.node)
+		return nil
+	}
+
+	if since := time.Since(ng.lastSent); since < time.Second*10 {
+		return fmt.Errorf("the node %q was synced less than %v ago, which is less than 10s", ng.node, since)
+	}
+
+	err := client.Sync(ctx, ng.node, zone, curGSAs)
+	if err == nil {
+		klog.Infof("On node %v, sent GSAs: %v", ng.node, curGSAs)
+		ng.lastGSAs = curGSAs
+		ng.lastSent = time.Now()
+	}
+	return err
+}
+
+type nodeZones struct {
+	sync.Mutex
+	m      map[string]string
+	client clientset.Interface
+}
+
+func newNodeZones(client clientset.Interface) *nodeZones {
+	return &nodeZones{
+		m:      make(map[string]string),
+		client: client,
+	}
+}
+
+func (nz *nodeZones) zoneByNode(nodeName string) (string, error) {
+	nz.Lock()
+	defer nz.Unlock()
+	if zone, ok := nz.m[nodeName]; ok {
+		return zone, nil
+	}
+	node, err := nz.client.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get node object: %w", err)
+	}
+	_, zone, _, err := parseNodeURL(node.Spec.ProviderID)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse node url: %w", err)
+	}
+	nz.m[nodeName] = zone
+	return zone, nil
+}
+
+// Copied from node_annotator.go
+func parseNodeURL(nodeURL string) (project, zone, instance string, err error) {
```

Comment:
GKE probably puts a label or annotation on the node, instead of having to hack apart this URL.